HL7-DaVinci/CDS-Library
# Implement Use of Shared CQL Libraries Across Rulesets

Kind of a lot going on here, but here are the major changes:

**1. Allowed access to `CDS_Connect_Commons_for_FHIRv400` and `DTRHelpers` shared CQL libraries across all rulesets:**
- Create references to each helper library within each ruleset's FHIR Library
- Include the helper libraries in each ruleset's prepopulation file.

**2. Implemented use of `CDS_Connect_Commons_for_FHIRv400` CQL library across all rulesets:**
- Added `CDS.` reference to CDS Library in front of CDS Connect Commons functions and removed the copy/pasted CDS Connect Commons functions from the bottom of each rulesets
- Some functions we were previously defining actually had similar functions within the updated CDS Library so I did some refactoring of function calls to standardize all of the rulesets with these CDS functions.
- It turns out a majority of the functions we use in DRLS are included within the CDS Connect Commons Library.

**3. Built out DTRHelpers shared CQL Library:**
- These are all of the DRLS functions that are not included within FHIRHelpers or CDS Connect Commons libraries.
- Most of them have to do with formatting the way we output data onto the questionnaire.
- No more functions are defined within the individual prepopulation files - all functions live within one of the shared libraries.

Here is a short explanation of why each function was included in DTRHelpers instead of using a function from CDS Connect:

![image](https://user-images.githubusercontent.com/51176978/91590481-80ad6680-e929-11ea-9b97-f825690fa0a6.png)

**4. Updates to  `CQLCookbook.md`:**
- Updated CQL Cookbook's "CQL Template" section to mention the shared FHIR libraries 
- Modified the Statement Templates to remove all of the function definitions, which are now abstracted to the shared libraries
- As a result, this instructional document is a lot cleaner and easier to follow for people who are new to writing CQL.


**Testing:**

I just tested all of the rulesets and they seem to work (except for Immunosuppressive drugs, which seems to be broken due to other reasons). I think the best way to test this will be:
- Ensure that questionnaire forms for each ruleset load and that all fields prepopulate the exact same was as they do for master branch.
- Take a look at the updated DTRHelpers Library (see function descriptions above).
- Make sure that all of the info within the CQL Cookbook still relevant and reflects these updates.
